src/DeltaLake/Kernel/Core/Table.cs

```diff
@@ -0,0 +1,356 @@
+                this.storageOptionsValueHandles = new GCHandle[count];
+                this.gcPinnedStorageOptionsKeyPtrs = (sbyte**)Marshal.AllocHGlobal(count * sizeof(sbyte*));
+                this.gcPinnedStorageOptionsValuePtrs = (sbyte**)Marshal.AllocHGlobal(count * sizeof(sbyte*));
+                this.storageOptionsKeySlices = new KernelStringSlice[count];
+                this.storageOptionsValueSlices = new KernelStringSlice[count];
+
+                foreach (KeyValuePair<string, string> kvp in tableStorageOptions.StorageOptions)
+                {
+                    (GCHandle keyHandle, IntPtr keyPtr) = kvp.Key.ToPinnedSBytePointer();
+                    (GCHandle valueHandle, IntPtr valuePtr) = kvp.Value.ToPinnedSBytePointer();
+
+                    this.storageOptionsKeyHandles[index] = keyHandle;
+                    this.storageOptionsValueHandles[index] = valueHandle;
+                    this.gcPinnedStorageOptionsKeyPtrs[index] = (sbyte*)keyPtr.ToPointer();
+                    this.gcPinnedStorageOptionsValuePtrs[index] = (sbyte*)valuePtr.ToPointer();
+                    this.storageOptionsKeySlices[index] = new KernelStringSlice { ptr = this.gcPinnedStorageOptionsKeyPtrs[index], len = (nuint)kvp.Key.Length };
+                    this.storageOptionsValueSlices[index] = new KernelStringSlice { ptr = this.gcPinnedStorageOptionsValuePtrs[index], len = (nuint)kvp.Value.Length };
+
+                    Methods.set_builder_option(this.kernelOwnedEngineBuilderPtr, this.storageOptionsKeySlices[index], this.storageOptionsValueSlices[index]);
+
+                    index++;
+                }
+
+                this.sharedExternEngine = Methods.builder_build(this.kernelOwnedEngineBuilderPtr);
+                if (this.sharedExternEngine.tag != ExternResultHandleSharedExternEngine_Tag.OkHandleSharedExternEngine)
+                {
+                    throw new InvalidOperationException("Could not build engine from the engine builder sent to Delta Kernel.");
+                }
+                this.kernelOwnedSharedExternEnginePtr = this.sharedExternEngine.Anonymous.Anonymous1.ok;
+                this.state = new ManagedTableState(this.tableLocationSlice, this.kernelOwnedSharedExternEnginePtr);
+                this.isKernelAllocated = true;
+            }
+        }
+
+        #region Delta Kernel table operations
+
+        internal Apache.Arrow.Table ReadAsArrowTable()
+        {
+            this.ThrowIfKernelNotSupported();
+
+            unsafe
+            {
+                return this.state.ArrowContext(true)->ToTable();
+            }
+        }
+
+        internal DataFrame ReadAsDataFrame()
+        {
+            this.ThrowIfKernelNotSupported();
+
+            unsafe
+            {
+#pragma warning disable CA2000 // DataFrames use the RecordBatch, so we don't need to dispose of it
+                return DataFrame.FromArrowRecordBatch(this.state.ArrowContext(true)->ToRecordBatch());

```

Comment:
Is using `ToRecordBatch` for efficient than enumerating the record batches, calling `DataFrame.FromArrowRecordBatch` on each and concatenating them with `DataFrame.Append(other.Rows)`?
I'm inclined to believe creating the large `RecordBatch` might be the way since `Append`, even if it takes a row collection still only does a record at a time.

Reply:
Ah so I originally did try out the ` DataFrame.Append(other.Rows)` approach, but I was getting a:

`System.NotImplementedexception "dictionary" DataFrame` within the DataFrame library. 

It seems like the `Rows` isn't able to parse into the DataFrame (but I didn't dig too much into the ML.NET code to understand why).

The only working implementation of `RecordBatch` to `DataFrame` I found was this GitHub issue:

https://github.com/apache/arrow/issues/35371

And incidentally, that author faced the same issue too:

![image](https://github.com/user-attachments/assets/aa712d3b-80c2-4d0a-b25c-5028dc22553e)

They got it working with `ArrowArrayConcatenator` - which is what I ended up implementing in `ToRecordBatch`.

But you're absolutely right, on a large dataset, this is definitely not the most efficient approach.

My thought process was, most folks would probably be doing this `DataFrame` interactivity for smaller datasets when doing development debugging (rather than pointing it at a Prod dataset at runtime), so just wanted to expose the `DataFrame` method as a friendly entry point.